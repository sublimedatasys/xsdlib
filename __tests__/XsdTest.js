import fs from "fs";
const { xsd2jsonSchema, xml2json } = require('../')

const readModuleFile = (
  path,
  callback
) => {
  try {
    const filename = require.resolve(path);
    fs.readFile(filename, "utf8", callback);
  } catch (e) {
    callback(e, "");
  }
};

it("Parse complex content xml2json", (done) => {
  readModuleFile("./XsdData.xsd", (_err, xsdText) => {
    const jsonO = xml2json(xsdText)
    expect(JSON.parse(jsonO)).toMatchSnapshot();
    done();
  });
});

it("Parse xsd restrictions", () => {
  const jsonO = xsd2jsonSchema(`
  <?xml version="1.0" encoding="UTF-8"?>
  <xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified" attributeFormDefault="unqualified">
    <xs:element name="TestElemet" type="GhbdtnId_Struct">
      <xs:annotation>
        <xs:documentation>Complex test element</xs:documentation>
      </xs:annotation>
    </xs:element>
	<xs:complexType name="GhbdtnId_Struct" final="#all">
		<xs:annotation>
			<xs:documentation>Ghwpa kwdfp  ghwp</xs:documentation>
		</xs:annotation>
		<xs:sequence>
			<xs:element name="Hps" type="Long_Type">
				<xs:annotation>
					<xs:documentation>Id kwdfp</xs:documentation>
				</xs:annotation>
			</xs:element>
			<xs:element name="Unc" type="SomeIdVariation48_Type">
				<xs:annotation>
					<xs:documentation>Id sfdwg</xs:documentation>
				</xs:annotation>
			</xs:element>
			<xs:element name="Xfw" type="Date_Type">
				<xs:annotation>
					<xs:documentation>Sdfs dghfd</xs:documentation>
				</xs:annotation>
			</xs:element>
			<xs:element name="SFwsdf" type="String4000_Type">
				<xs:annotation>
					<xs:documentation>sdfsd sdfsdf</xs:documentation>
				</xs:annotation>
			</xs:element>
			<xs:element name="SFwsdf2">
				<xs:annotation>
					<xs:documentation>sdfsd sdfsdf</xs:documentation>
				</xs:annotation>
        <xs:simpleType>
          <xs:restriction base="xs:string">
            <xs:maxLength value="5000"/>
          </xs:restriction>
        </xs:simpleType>
			</xs:element>
		</xs:sequence>
	</xs:complexType>
	<xs:simpleType name="Long_Type">
		<xs:annotation>
			<xs:documentation/>
		</xs:annotation>
		<xs:restriction base="xs:long"/>
	</xs:simpleType>
	<xs:simpleType name="SomeIdVariation48_Type">
		<xs:restriction base="xs:string">
			<xs:pattern value="([0-9]|[A-F]){48}"/>
		</xs:restriction>
	</xs:simpleType>
	<xs:simpleType name="Date_Type">
		<xs:annotation>
			<xs:documentation>Date</xs:documentation>
		</xs:annotation>
		<xs:restriction base="xs:date"/>
	</xs:simpleType>
	<xs:simpleType name="String4000_Type">
		<xs:restriction base="xs:string">
			<xs:maxLength value="4000"/>
		</xs:restriction>
	</xs:simpleType>
  </xs:schema>
`)
  expect(JSON.parse(jsonO)).toMatchSnapshot();
});

it("Parse xsd extention", () => {
  const jsonO = xsd2jsonSchema(`
  <?xml version="1.0" encoding="UTF-8"?>
  <xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified" attributeFormDefault="unqualified">
    <xs:element name="TestElemet" type="BDtPrasd">
      <xs:annotation>
        <xs:documentation>Complex test element</xs:documentation>
      </xs:annotation>
    </xs:element>

    <xs:complexType name="BDtPrasd">
		<xs:annotation>
			<xs:documentation>Lorem ipsum dolor sit amet, consectetur adipisicing elit. Doloremque perspiciatis possimus corrupti laboriosam reiciendis repellendus iure? Cum recusandae distinctio officiis ipsa ratione voluptatum eum provident officia, fugiat, maiores esse nesciunt.</xs:documentation>
		</xs:annotation>
		<xs:complexContent>
			<xs:extension base="BDtBase">
				<xs:sequence>
					<xs:element name="Fsap" type="String64_Type" minOccurs="0">
						<xs:annotation>
							<xs:documentation>Lorem ipsum dolor, sit amet consectetur adipisicing elit. Dolores cupiditate delectus atque quae, error, sequi at quas unde earum provident expedita recusandae assumenda sint corrupti odio laboriosam. Saepe, voluptatem laborum!</xs:documentation>
						</xs:annotation>
					</xs:element>
					<xs:element name="DateGFwp" type="Date_Type" minOccurs="0">
						<xs:annotation>
							<xs:documentation>Lorem ipsum dolor, sit amet consectetur adipisicing elit. Odit impedit cumque accusamus porro! Quo, nisi corporis obcaecati temporibus unde quisquam odit dignissimos! Voluptates sequi placeat eveniet perferendis repellat cumque numquam!</xs:documentation>
						</xs:annotation>
					</xs:element>
					<xs:element name="Npwfd" type="String4000_Type" minOccurs="0">
						<xs:annotation>
							<xs:documentation>Lorem, ipsum dolor sit amet consectetur adipisicing elit. Numquam, voluptates voluptatem eius mollitia cupiditate odio, nemo qui dolorum placeat alias sunt. Nam non voluptatibus, itaque magni similique saepe! Maxime, nisi.</xs:documentation>
						</xs:annotation>
					</xs:element>
				</xs:sequence>
			</xs:extension>
		</xs:complexContent>
	</xs:complexType>
  <xs:complexType name="BDtBase" abstract="true">
		<xs:annotation>
			<xs:documentation>Base</xs:documentation>
		</xs:annotation>
		<xs:sequence>
			<xs:element name="Prasd" type="Prasd_Struct">
				<xs:annotation>
					<xs:documentation>Prasd</xs:documentation>
				</xs:annotation>
			</xs:element>
		</xs:sequence>
	</xs:complexType>
	<xs:complexType name="Prasd_Struct">
		<xs:annotation>
			<xs:documentation>Prasd asdfa</xs:documentation>
		</xs:annotation>
		<xs:sequence>
			<xs:element name="Apwoe" type="Max100_Type" minOccurs="0">
				<xs:annotation>
					<xs:documentation>Prasd dfe</xs:documentation>
				</xs:annotation>
			</xs:element>
			<xs:element name="Ghbdtn" type="SomeIdVariation48_Type" minOccurs="0">
				<xs:annotation>
					<xs:documentation>Prasd fgher</xs:documentation>
				</xs:annotation>
			</xs:element>
		</xs:sequence>
  </xs:complexType>

	<xs:simpleType name="String64_Type">
		<xs:restriction base="xs:string">
			<xs:maxLength value="64"/>
		</xs:restriction>
	</xs:simpleType>
  <xs:simpleType  name="Max100_Type">
  <xs:restriction base="xs:integer">
    <xs:minInclusive value="0"/>
    <xs:maxInclusive value="100"/>
  </xs:restriction>
</xs:simpleType>
	<xs:simpleType name="Long_Type">
		<xs:annotation>
			<xs:documentation/>
		</xs:annotation>
		<xs:restriction base="xs:long"/>
	</xs:simpleType>
	<xs:simpleType name="SomeIdVariation48_Type">
		<xs:restriction base="xs:string">
			<xs:pattern value="([0-9]|[A-F]){48}"/>
		</xs:restriction>
	</xs:simpleType>
	<xs:simpleType name="Date_Type">
		<xs:annotation>
			<xs:documentation>Date</xs:documentation>
		</xs:annotation>
		<xs:restriction base="xs:date"/>
	</xs:simpleType>
	<xs:simpleType name="String4000_Type">
		<xs:restriction base="xs:string">
			<xs:maxLength value="4000"/>
		</xs:restriction>
	</xs:simpleType>
  </xs:schema>
`)
  expect(JSON.parse(jsonO)).toMatchSnapshot();
});

it("Parse xsd array", () => {
  const jsonO = xsd2jsonSchema(`

  <?xml version="1.0" encoding="UTF-8"?>
  <xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified" attributeFormDefault="unqualified">
    <xs:element name="TestElemet" type="A1_Struct">
    </xs:element>

  <xs:complexType name="A1_Struct">
		<xs:complexContent>
			<xs:extension base="RSt_Struct">
				<xs:sequence>
					<xs:element name="Dt" type="Date_Type">
					</xs:element>
					<xs:element name="ChSrv" type="Ch_struct" minOccurs="0" maxOccurs="unbounded">
					</xs:element>
					<xs:element name="Dt2" type="Date_Type">
					</xs:element>
					<xs:element name="Ch2" type="Ch_struct2">
					</xs:element>
				</xs:sequence>
			</xs:extension>
		</xs:complexContent>
	</xs:complexType>
	<xs:complexType name="RSt_Struct">
		<xs:sequence>
			<xs:element name="Rdg" type="String64_Type">
			</xs:element>
			<xs:element name="Rdt" type="Max100_Type">
			</xs:element>
		</xs:sequence>
	</xs:complexType>
	<xs:complexType name="Ch_struct">
		<xs:sequence>
			<xs:element name="Wdt" type="String4000_Type">
			</xs:element>
			<xs:element name="Srv" type="Long_Type">
			</xs:element>
		</xs:sequence>
	</xs:complexType>
	<xs:complexType name="Ch_struct2">
		<xs:sequence>
			<xs:element name="Wdt" type="String4000_Type">
			</xs:element>
			<xs:element name="Srv" type="Long_Type">
			</xs:element>
      <xs:element name="ChSrv" type="Ch_struct" minOccurs="0" maxOccurs="unbounded">
      </xs:element>
		</xs:sequence>
	</xs:complexType>

	<xs:simpleType name="String64_Type">
		<xs:restriction base="xs:string">
			<xs:maxLength value="64"/>
		</xs:restriction>
	</xs:simpleType>
  <xs:simpleType  name="Max100_Type">
  <xs:restriction base="xs:integer">
    <xs:minInclusive value="0"/>
    <xs:maxInclusive value="100"/>
  </xs:restriction>
</xs:simpleType>
	<xs:simpleType name="Long_Type">
		<xs:annotation>
			<xs:documentation/>
		</xs:annotation>
		<xs:restriction base="xs:long"/>
	</xs:simpleType>
	<xs:simpleType name="SomeIdVariation48_Type">
		<xs:restriction base="xs:string">
			<xs:pattern value="([0-9]|[A-F]){48}"/>
		</xs:restriction>
	</xs:simpleType>
	<xs:simpleType name="Date_Type">
		<xs:annotation>
			<xs:documentation>Date</xs:documentation>
		</xs:annotation>
		<xs:restriction base="xs:date"/>
	</xs:simpleType>
	<xs:simpleType name="String4000_Type">
		<xs:restriction base="xs:string">
			<xs:maxLength value="4000"/>
		</xs:restriction>
	</xs:simpleType>
  </xs:schema>
`)
  expect(JSON.parse(jsonO)).toMatchSnapshot();
});

it("Parse complex content xsd2jsonSchema", (done) => {
  readModuleFile("./XsdData.xsd", (err, xsdText) => {
    const jsonO = xsd2jsonSchema(xsdText)
    expect(JSON.parse(jsonO)).toMatchSnapshot();
    done();
  });
});

it('Parse xsd has no complex types', () => {
	const jsonO = xsd2jsonSchema(`
  	<?xml version="1.0" encoding="UTF-8"?>
	<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified" attributeFormDefault="unqualified">
	  <xs:element name="TestElemet" type="SomeIdVariation48_Type">
	  </xs:element>
	  <xs:simpleType name="SomeIdVariation48_Type">
		  <xs:restriction base="xs:string">
			  <xs:pattern value="[0-9]{4}"/>
		  </xs:restriction>
	  </xs:simpleType>
	</xs:schema>
  `)
	expect(JSON.parse(jsonO)).toMatchSnapshot();
});

it('Parse xsd pattern extra escaping characters dot', () => {
	  const jsonO = xsd2jsonSchema(`
		<?xml version="1.0" encoding="UTF-8"?>
	  <xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified" attributeFormDefault="unqualified">
		<xs:element name="TestElemet" type="SomeIdVariation48_Type">
		</xs:element>
		<xs:complexType name="Ch_struct2">
			<xs:sequence>
				<xs:element name="Wdt" type="xs:string">
				</xs:element>
			</xs:sequence>
		</xs:complexType>
		<xs:simpleType name="SomeIdVariation48_Type">
			<xs:restriction base="xs:string">
				<xs:pattern value="(\\.[0-9]{3})?"/>
			</xs:restriction>
		</xs:simpleType>
	  </xs:schema>
	`)
	  expect(JSON.parse(jsonO)).toMatchSnapshot();
});

it('Parse xsd pattern extra escaping characters plus', () => {
	  const jsonO = xsd2jsonSchema(`
		<?xml version="1.0" encoding="UTF-8"?>
	  <xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified" attributeFormDefault="unqualified">
		<xs:element name="TestElemet" type="SomeIdVariation48_Type">
		</xs:element>
		<xs:complexType name="Ch_struct2">
			<xs:sequence>
				<xs:element name="Wdt" type="xs:string">
				</xs:element>
			</xs:sequence>
		</xs:complexType>
		<xs:simpleType name="SomeIdVariation48_Type">
			<xs:restriction base="xs:string">
				<xs:pattern value="((-|\\+)[0-9]{2}:[0-9]{2})?)?)?"/>
			</xs:restriction>
		</xs:simpleType>
	  </xs:schema>
	`)
	  expect(JSON.parse(jsonO)).toMatchSnapshot();
});
