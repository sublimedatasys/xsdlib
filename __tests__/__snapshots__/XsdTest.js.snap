// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parse complex content xml2json 1`] = `
Object {
  "properties": Object {
    "xs:schema": Object {
      "attributeFormDefault": "unqualified",
      "elementFormDefault": "qualified",
      "properties": Object {
        "xs:complexType": Object {
          "items": Object {
            "name": "TestElement_Struct",
            "properties": Object {
              "xs:annotation": Object {
                "properties": Object {
                  "xs:documentation": Object {
                    "default": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Vero ipsa sunt commodi sed minus fuga, cupiditate nobis laudantium? Maxime autem perspiciatis vero neque quos sit, sequi recusandae veniam suscipit saepe. dfghje",
                    "type": "string",
                  },
                },
                "type": "object",
              },
              "xs:complexContent": Object {
                "properties": Object {
                  "xs:extension": Object {
                    "base": "Middle_Struct",
                    "properties": Object {
                      "xs:sequence": Object {
                        "properties": Object {
                          "xs:element": Object {
                            "name": "BDt",
                            "properties": Object {
                              "xs:complexType": Object {
                                "properties": Object {
                                  "xs:complexContent": Object {
                                    "properties": Object {
                                      "xs:restriction": Object {
                                        "base": "BDtLewfPsd",
                                        "properties": Object {
                                          "xs:sequence": Object {
                                            "properties": Object {
                                              "xs:sequence": Object {
                                                "items": Object {
                                                  "properties": Object {
                                                    "xs:element": Object {
                                                      "name": "Rfvf",
                                                      "properties": Object {
                                                        "xs:annotation": Object {
                                                          "properties": Object {
                                                            "xs:documentation": Object {
                                                              "default": "Lorem ipsum dolor sit amet consectetur, adipisicing elit. In sint hic dignissimos sapiente voluptatem autem ex ratione consequatur debitis ut rerum iste magnam, amet incidunt. Officia exercitationem accusantium amet molestias.",
                                                              "type": "string",
                                                            },
                                                          },
                                                          "type": "object",
                                                        },
                                                      },
                                                      "type": "ApwoeClass_Type",
                                                    },
                                                  },
                                                  "type": "object",
                                                },
                                                "type": "array",
                                              },
                                            },
                                            "type": "object",
                                          },
                                        },
                                        "type": "object",
                                      },
                                    },
                                    "type": "object",
                                  },
                                },
                                "type": "object",
                              },
                            },
                            "type": "object",
                          },
                        },
                        "type": "object",
                      },
                    },
                    "type": "object",
                  },
                },
                "type": "object",
              },
            },
            "type": "object",
          },
          "type": "array",
        },
        "xs:element": Object {
          "name": "TestElemet",
          "properties": Object {
            "xs:annotation": Object {
              "properties": Object {
                "xs:documentation": Object {
                  "default": "Complex test element",
                  "type": "string",
                },
              },
              "type": "object",
            },
          },
          "type": "TestElement_Struct",
        },
        "xs:simpleType": Object {
          "items": Object {
            "name": "DefsId_Type",
            "properties": Object {
              "xs:annotation": Object {
                "properties": Object {
                  "xs:documentation": Object {
                    "default": "Lorem ipsum dolor sit amet consectetur adipisicing elit. Minus delectus aliquam corporis quos molestiae quasi facilis eum accusantium sed quaerat nisi, odit nostrum dolor, aspernatur deserunt, voluptates similique tenetur architecto!",
                    "type": "string",
                  },
                },
                "type": "object",
              },
              "xs:restriction": Object {
                "base": "xs:string",
                "properties": Object {
                  "xs:maxLength": Object {
                    "properties": Object {},
                    "type": "object",
                    "value": "3",
                  },
                },
                "type": "object",
              },
            },
            "type": "object",
          },
          "type": "array",
        },
      },
      "type": "object",
      "xmlns:xs": "http://www.w3.org/2001/XMLSchema",
    },
  },
  "type": "object",
}
`;

exports[`Parse complex content xsd2jsonSchema 1`] = `
Object {
  "properties": Object {
    "TestElemet": Object {
      "properties": Object {
        "ComplexField2": Object {
          "properties": Object {
            "DefsId": Object {
              "maxLength": "3",
              "type": "string",
            },
            "DepartId": Object {
              "type": "long",
            },
            "ShortCodeField": Object {
              "maxLength": "4",
              "type": "string",
            },
          },
          "type": "object",
        },
        "MidCketysqDate": Object {
          "type": "date",
        },
        "MidCketysqQPwo": Object {
          "properties": Object {
            "Alks": Object {
              "type": "long",
            },
            "Eiwo": Object {
              "type": "long",
            },
            "Ruty": Object {
              "type": "long",
            },
          },
          "type": "object",
        },
        "MidCnfhibqQPwo": Object {
          "properties": Object {
            "Alks": Object {
              "type": "long",
            },
            "Eiwo": Object {
              "type": "long",
            },
            "Ruty": Object {
              "type": "long",
            },
          },
          "type": "object",
        },
        "MidLtncrbqQPwo": Object {
          "items": Object {
            "properties": Object {
              "Gjljlb": Object {
                "type": "boolean",
              },
              "Surei": Object {
                "properties": Object {
                  "Alks": Object {
                    "type": "long",
                  },
                  "Eiwo": Object {
                    "type": "long",
                  },
                  "Ruty": Object {
                    "type": "long",
                  },
                },
                "type": "object",
              },
            },
            "type": "object",
          },
          "type": "array",
        },
        "SomeId": Object {
          "maxLength": "36",
          "type": "string",
        },
        "SomeIdVariation": Object {
          "pattern": "([0-9]|[A-F]){48}",
          "type": "string",
        },
        "SomeName": Object {
          "maxLength": "36",
          "type": "string",
        },
      },
      "type": "object",
    },
  },
  "type": "object",
}
`;

exports[`Parse xsd array 1`] = `
Object {
  "properties": Object {
    "TestElemet": Object {
      "properties": Object {
        "Ch2": Object {
          "properties": Object {
            "ChSrv": Object {
              "items": Object {
                "properties": Object {
                  "Srv": Object {
                    "type": "long",
                  },
                  "Wdt": Object {
                    "maxLength": "4000",
                    "type": "string",
                  },
                },
                "type": "object",
              },
              "type": "array",
            },
            "Srv": Object {
              "type": "long",
            },
            "Wdt": Object {
              "maxLength": "4000",
              "type": "string",
            },
          },
          "type": "object",
        },
        "ChSrv": Object {
          "items": Object {
            "properties": Object {
              "Srv": Object {
                "type": "long",
              },
              "Wdt": Object {
                "maxLength": "4000",
                "type": "string",
              },
            },
            "type": "object",
          },
          "type": "array",
        },
        "Dt": Object {
          "type": "date",
        },
        "Dt2": Object {
          "type": "date",
        },
        "Rdg": Object {
          "maxLength": "64",
          "type": "string",
        },
        "Rdt": Object {
          "maximum": 100,
          "minimum": 0,
          "type": "integer",
        },
      },
      "type": "object",
    },
  },
  "type": "object",
}
`;

exports[`Parse xsd extention 1`] = `
Object {
  "properties": Object {
    "TestElemet": Object {
      "properties": Object {
        "DateGFwp": Object {
          "type": "date",
        },
        "Fsap": Object {
          "maxLength": "64",
          "type": "string",
        },
        "Npwfd": Object {
          "maxLength": "4000",
          "type": "string",
        },
        "Prasd": Object {
          "properties": Object {
            "Apwoe": Object {
              "maximum": 100,
              "minimum": 0,
              "type": "integer",
            },
            "Ghbdtn": Object {
              "pattern": "([0-9]|[A-F]){48}",
              "type": "string",
            },
          },
          "type": "object",
        },
      },
      "type": "object",
    },
  },
  "type": "object",
}
`;

exports[`Parse xsd has no complex types 1`] = `
Object {
  "properties": Object {
    "TestElemet": Object {
      "pattern": "[0-9]{4}",
      "type": "string",
    },
  },
  "type": "object",
}
`;

exports[`Parse xsd pattern extra escaping characters dot 1`] = `
Object {
  "properties": Object {
    "TestElemet": Object {
      "pattern": "(.[0-9]{3})?",
      "type": "string",
    },
  },
  "type": "object",
}
`;

exports[`Parse xsd pattern extra escaping characters plus 1`] = `
Object {
  "properties": Object {
    "TestElemet": Object {
      "pattern": "((-|+)[0-9]{2}:[0-9]{2})?)?)?",
      "type": "string",
    },
  },
  "type": "object",
}
`;

exports[`Parse xsd restrictions 1`] = `
Object {
  "properties": Object {
    "TestElemet": Object {
      "properties": Object {
        "Hps": Object {
          "type": "long",
        },
        "SFwsdf": Object {
          "maxLength": "4000",
          "type": "string",
        },
        "SFwsdf2": Object {
          "maxLength": "5000",
          "type": "string",
        },
        "Unc": Object {
          "pattern": "([0-9]|[A-F]){48}",
          "type": "string",
        },
        "Xfw": Object {
          "type": "date",
        },
      },
      "type": "object",
    },
  },
  "type": "object",
}
`;
